includes/class-wcsr-resource.php

```diff
@@ -250,6 +250,26 @@ public function get_days_active( $from_timestamp, $to_timestamp = null ) {
 		return $days_active;
 	}
 
+	/**
+	 * Calculates the number of whole (i.e. using floor) days between two timestamps.
+	 * Uses the same logic in @see wcs_estimate_periods_between() but this function exclusively calculates the number of days
+	 *
+	 * @since 1.1.0
+	 * @param int $start_timestamp The starting timestamp to calculate the number of days from
+	 * @param int $end_timestamp   The end timestamp to calculate the number of days to
+	 * @return int
+	 */
+	public function get_days_in_period( $start_timestamp, $end_timestamp ) {
+		// return 0 if the start timestamp is after the end timestamp
+		$periods_until = 0;
+
+		if ( $end_timestamp > $start_timestamp ) {
+			$periods_until = floor( ( $end_timestamp - $start_timestamp ) / DAY_IN_SECONDS );
+		}
+
+		return $periods_until;
```

Comment:
Why does this method live within `WCSR_Resource`? It doesn't use `$this` at all, and instead does an operation based on two timestamps passed it. Seems more like it belongs in an independent utility function than belonging to each instance of a resource.

Reply:
Yeah okay i can see a few more of the new functions that fall into this same category which have been added recently to the `WCSR_Resource` class and would probably fit better in a separate utility file.. ie.

 - `get_days_in_period()`
 - `get_active_days_ratio()`
 - `is_on_same_day()`
 - `get_timestamps_between()` is also another that could be moved but it's been in the class since the start)

Would you be okay with all of these being moved out of WCSR resources class and into a separate `wcsr-time-functions.php` file or or maybe named something different?

Reply:
> i can see a few more of the new functions that fall into this same category  

Good finds. 👍 

> Would you be okay with

Yep for sure. Sounds like they all belong as utility functions. A `WCSR_Resource` really should only have methods that either need access to that instances private data and/or act in some other way that may differ between instances.

Anything dealing with data passed directly into it, especially if it's generic data like date/time data, and wouldn't need to have its behaviour modified for different instances of a resources (including child classes), is better in a separate function in a file like `wcsr-time-functions.php`.

Date/time data is a really good candidate for separate API functions, because no matter what class is using `get_timestamps_between()`, it's always going to return the same value (hence why it's `static`) because time is (sort of) a universal truth.

But the same could apply also 

> i remember your comment saying this should be changed to a protected method 

I think having it as a separate file, even though that makes it public, makes more sense than making it `protected` if it doesn't need to be on the instance of the resource.

> i think this function can stay put for now as it does kinda related directly to the resource

I think it should be moved. It's `protected`, so is pretty safe to move (especially given AFAIK, no one else is using this library yet). It's also marked as `static`, which is a smell to signify that it really doesn't belong in the resource instances.

A small piece of proof that these utility functions belong outside `WCSR_Resource` comes from unit testing. Once `get_timestamps_between()` is a separate function, there will be no reason to instantiate a mock `WCSR_Resource`. It's pretty clear we don't need a mock `WCSR_Resource` to test the function, because we wouldn't need to stub any methods or set any mock data on it, which makes it pretty clear it doesn't belong on each instance of a resource.

Reply:
Thanks @thenbrent! That all make sense and definitely makes `WCSR_Resource` feel a lot clearner.

As per the above comment and also the comment in the [slack thread](https://prospress.slack.com/archives/C1QQF18V7/p1511372973000681?thread_ts=1511244102.000087&cid=C1QQF18V7), I created a new function to get the line item name. It didn't make sense for that to live in the Resource manager class so i created a  new file `wcsr-renewal-functions.php`. With that, i also move the creating line item function to this new file as well because it really didn't need to live in manager class and i felt it belonged to the same category as the other protected functions you mentioned above.

Let me know what you think with the [recent commits](https://github.com/Prospress/woocommerce-subscriptions-resource/compare/1c279c2...7e823d2) (hoping i didn't get too far with the separation of functions) :)